src/main/java/org/tron/common/runtime/vm/program/invoke/ProgramInvokeImpl.java

```diff
@@ -19,296 +19,362 @@
-        this.caller = caller;
-        this.balance = balance;
-        this.callValue = callValue;
-        this.msgData = msgData;
-
-        // last Block env
-        this.prevHash = lastHash;
-        this.coinbase = coinbase;
-        this.timestamp = timestamp;
-        this.number = number;
-
-        this.deposit = deposit;
-        this.byTransaction = false;
-        this.isStaticCall = isStaticCall;
-        this.byTestingSuite = byTestingSuite;
-
-        this.dropLimit = balance.clone();
+  // private BlockStore blockStore;
+  /* TRANSACTION  env*/
+  private final DataWord address;
+  private final DataWord origin, caller, balance, callValue;
+  private byte[] msgData;
+
+  private long vmStartInUs;
+  private long vmShouldEndInUs;
+
+  /* BLOCK  env **/
+  private final DataWord prevHash, coinbase, timestamp, number;
+
+  private Deposit deposit = null;
+  private boolean byTransaction = true;
+  private boolean byTestingSuite = false;
+  private int callDeep = 0;
+  private boolean isStaticCall = false;
+
+  public ProgramInvokeImpl(DataWord address, DataWord origin, DataWord caller, DataWord balance,
+      DataWord callValue, byte[] msgData,
+      DataWord lastHash, DataWord coinbase, DataWord timestamp, DataWord number,
+      DataWord difficulty,
+      Deposit deposit, int callDeep, boolean isStaticCall, boolean byTestingSuite,
+      long vmStartInUs, long vmShouldEndInUs) {
+    this.address = address;
+    this.origin = origin;
+    this.caller = caller;
+    this.balance = balance;
+    this.callValue = callValue;
+    this.msgData = msgData;
+
+    // last Block env
+    this.prevHash = lastHash;
+    this.coinbase = coinbase;
+    this.timestamp = timestamp;
+    this.number = number;
+
+    this.deposit = deposit;
+    this.byTransaction = false;
+    this.isStaticCall = isStaticCall;
+    this.byTestingSuite = byTestingSuite;
+    this.vmStartInUs = vmStartInUs;
+    this.vmShouldEndInUs = vmShouldEndInUs;
+
+  }
+
+  public ProgramInvokeImpl(byte[] address, byte[] origin, byte[] caller, long balance,
+      byte[] callValue, byte[] msgData,
+      byte[] lastHash, byte[] coinbase, long timestamp, long number, Deposit deposit,
+      long vmStartInUs, long vmShouldEndInUs, boolean byTestingSuite) {
+    this(address, origin, caller, balance, callValue, msgData, lastHash, coinbase,
+        timestamp, number, deposit, vmStartInUs, vmShouldEndInUs);
+    this.byTestingSuite = byTestingSuite;
+  }
+
+  public ProgramInvokeImpl(byte[] address, byte[] origin, byte[] caller, long balance,
+      byte[] callValue, byte[] msgData, byte[] lastHash, byte[] coinbase, long timestamp,
+      long number, Deposit deposit, long vmStartInUs, long vmShouldEndInUs) {
+
+    // Transaction env
+    this.address = new DataWord(address);
+    this.origin = new DataWord(origin);
+    this.caller = new DataWord(caller);
+    this.balance = new DataWord(balance);
+    this.callValue = new DataWord(callValue);
+    this.msgData = msgData;
```

Comment:
![Vulnerability](https://sonarsource.github.io/sonar-github/vulnerability.png 'Vulnerability') Vulnerability: Store a copy of "msgData". ([squid:S2384](https://sonarcloud.io/organizations/tron-zhaohong/rules#rule_key=squid%3AS2384))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=java-tron&pullRequest=1188&issues=AWTf_bHJrFWWQMKkhNxB&open=AWTf_bHJrFWWQMKkhNxB)

[ISSUE_KEY]: AWTf_bHJrFWWQMKkhNxB